daphne/src/testing.rs

```diff
@@ -467,37 +467,104 @@ impl AuditLog for MockAuditLog {
     }
 }
 
-#[derive(Default)]
+/// Aggregate share and associated book-keeping data for a bucket of reports.
+//
+// TODO heavy hitters: Move this code into a space where `daphne_server` can use it. We need the
+// same logic there.
 #[cfg_attr(any(test, feature = "test-utils"), derive(deepsize::DeepSizeOf))]
-pub(crate) struct AggregateStoreForCollection {
-    /// The aggregate share.
-    pub(crate) agg_share: DapAggregateShare,
-    /// Indication of whether the aggregate share has been collected.
-    pub(crate) collected: bool,
-    /// The reports included in the aggregate share.
-    pub(crate) reports: HashSet<ReportId>,
+pub struct AggregateStore {
+    pub agg_share: DapAggregateShare,
+
+    /// Indication of whether the bucket has been collected at least once. If set, new reports for
+    /// this bucket will be rejected.
+    pub collected: bool,
+
+    /// The reports included in the current aggregate share. If a report wants to be aggregated is
+    /// already in this set, it will be rejected.
+    pub reports: HashSet<ReportId>,
+
+    /// The current aggregation parameter.
+    pub(crate) agg_param: DapAggregationParam,
+}
+
+impl AggregateStore {
+    /// Check if the given aggregation parameter can be used to aggregate the reports in this bucket.
+    ///
+    /// For IDPF-based VDAFs (e.g., Mastic), this returns true only if the aggregation parameter
+    /// results in a valid tree traversal, that is, each prefix encoded by the aggregation
+    /// parameter is equal to or a child of a prefix in the current aggregation parameter.
+    ///
+    /// For simple VDAFs (e.g., Prio3), only the empty aggregation parameter is valid.
+    //
+    // TODO heavy hitters: Enforce the maximum tree depth.
+    pub fn is_valid(&self, next_agg_param: &DapAggregationParam) -> bool {
```

Comment:
Is there any reason for this method to be pub? I don't like methods that return `bool` in public APIs as they tend to be misused and distract from the most appropriate methods (in this case goto).

Another point is that maybe this function should instead return `Result<(), DapAbort>` so that `goto` can return `Result<&mut Self, DapAbort>` so we don't lose the information of why you couldn't progress

Reply:
No reason to be public, removed.

Propagating an error might be useful.  A `DapAbort` isn't appropriate because the caller wouldn't want to handle this generically as an abort. The correct action depends on the context. For instance, if this error occurs while processing an aggregation job, then the reports for this bucket should be rejected, rather than abort the aggregation job altogether.

Also, in the rejection case, we won't inform the Leader why the aggregation parameter is bad, just that we rejected the report because of it. (This is an indication of a misbehaving Leader.)